### PR TITLE
Fix nats-operator

### PR DESCRIPTION
**Pull request by marcobebway**
**Description**

During kubernetes version upgrades, sometimes nats-operator could not create nats-server pods and remains in this state without self-healing. This pull request fixes this issue by starting a nats-operator-doctor app as an additional container in the Eventing controller, that observes the nats-cluster state and restarts the nats-operator if it detects that the latter fails to create the desired nats-server pods within a given period of time.

**Related issue(s)**

- Fixes https://github.com/kyma-project/kyma/issues/11439.

### CODE REVIEW

File: components/nats-operator/pkg/doctor/state.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	labelValNatsCluster = "eventing-nats"
+
+	// natsClusterName is used as a named key for NatsCluster CustomResource.
+	natsClusterName = "eventing-nats"
+	// natsOperatorDeploymentName is used as a value for NatsCluster CustomResource.
+	natsOperatorDeploymentName = "nats-operator"
+)
+
+// state represents the state of a NATS cluster.
+type state struct {
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/nats-operator/pkg/doctor/state.go:43]: The `NATS cluster state` and `NATS operator doctor` has two different concerns. The state is concerned with making sure that the NATS cluster state is always updated and present in the state object. On the other hand, the doctor is concerned with making sure that the NATS operator is doing its job correctly.

File: components/nats-operator/pkg/errors/errors.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+type recoverableError struct {
+	err error
+}
+
+// Recoverable returns a new recoverable error instance.
+func Recoverable(err error) error {
+	return &recoverableError{err: err}
+}
+
+// recover implements the recoverable interface.
```
[the remark this answers is not in the document]
Comment (marcobebway) [reply components/nats-operator/pkg/errors/errors.go:34]: It acts as a marker interface to distinguish between any kind of error and recoverable errors. The actual error recovering is delegated to the caller.

[the remark this answers is not in the document]
Comment (marcobebway) [reply components/nats-operator/pkg/errors/errors.go:34]: It acts like a marker interface to distinguish between any kind of error and recoverable errors, but the actual error recovering is delegated to the caller.

[the remark this answers is not in the document]
Comment (marcobebway) [reply components/nats-operator/pkg/errors/errors.go:34]: It acts like a marker interface to distinguish between any kind of errors and recoverable errors, but the actual error recovering is delegated to the caller.

**Review by radufa — APPROVED** [review]

**Pull request merged**